### PR TITLE
Send push information to trust code out-of-band

### PR DESCRIPTION
**Pull request by aaronlehmann**
The trust code used to parse the console output of `docker push` to
extract the digest, tag, and size information and determine what to
sign. This is fragile and might give an attacker control over what gets
signed if the attacker can find a way to influence what gets printed as
part of the push output.

This commit sends the push metadata out-of-band. It introduces an `Aux`
field in JSONMessage that can carry application-specific data alongside
progress updates. Instead of parsing formatted output, the client looks
in this field to get the digest, size, and tag from the push.

cc @diogomonica @dmcgowan

### CODE REVIEW

**Comment by diogomonica** [comment]
@docker/security-team

File: api/client/trust.go [hunk 1]
```diff
@@ -25,8 +22,8 @@ import (
 	"github.com/docker/docker/api/types"
 	registrytypes "github.com/docker/docker/api/types/registry"
 	"github.com/docker/docker/cliconfig"
-	"github.com/docker/docker/pkg/ansiescape"
```
Comment (dnephin) [inline api/client/trust.go]: I believe this is the last place `ansiescape` was used, and that package can be removed now.

Comment (aaronlehmann) [inline api/client/trust.go]: Nice! Added a commit to remove this package.

**Comment by dmcgowan** [comment]
LGTM

**Comment by diogomonica** [comment]
This looks good, thank you Aaron.

Question, should we have a signature specific field? I'm good either way, I just think that if it was a signature field people would be more careful when reviewing it/handling it than if it is called "Aux". Also I don't see a lot of other use-cases out of the top of my head.

LGTM regardless

**Comment by cyli** [comment]
LGTM

**Comment by aaronlehmann** [comment]
> Question, should we have a signature specific field? I'm good either way, I just think that if it was a signature field people would be more careful when reviewing it/handling it than if it is called "Aux". Also I don't see a lot of other use-cases out of the top of my head.

@diogomonica This is a good point. I was reluctant to couple the `JSONMessage` structure (which is already a complete mess) to trust specifically. Same goes for intermediate representations like the `Progress` structure.

I think the worst thing someone could do by mishandling the `Aux` field (without type assertions that mess with the actual payload) is prevent trusted push from working at all, which would break the tests. So I'm fairly comfortable doing it this way. But I'm open to alternatives if there are any ideas.

**Comment by aaronlehmann** [comment]
Rebased.

**Comment by dmcgowan** [comment]
after rebase LGTM

**Comment by thaJeztah** [comment]
ping @tonistiigi could you have a look at this one?

**Comment by aaronlehmann** [comment]
Rebased again.

**Comment by tonistiigi** [comment]
LGTM

**Comment by crosbymichael** [comment]
LGTM

**Pull request merged**